Document reference shape entity naming (#935)
Ferrite-FEM/Ferrite.jl: docs/src/topics/reference_shapes.md

# Reference shapes

The reference shapes in Ferrite are used to define grid cells,
function interpolations (i.e. shape functions), and quadrature rules.
Currently, the following reference shapes are defined

* `RefLine`
* `RefTriangle`
* `RefQuadrilateral`
* `RefTetrahedron`
* `RefHexahedron`
* `RefPrism`
* `RefPyramid`

## Entity naming
Ferrite denotes the entities of a reference shape as follows

| Entity   | Description |
| :------- | :---------- |

> | `Vertex` | 0-dimensional points in the reference shape. |
> | `Edge`   | 1-dimensional entities connecting two vertices. |
> | `Face`   | 2-dimensional entities whose boundary is defined by 3 or more edges. |
> | `Volume` | 3-dimensional entity enclosed by faces. |

Reviewer: ```suggestion
| Vertex | 0-dimensional entity in the reference shape. |
| Edge   | 1-dimensional entity connecting two vertices. |
| Face   | 2-dimensional entity enclosed by edges. |
| Volume | 3-dimensional entity enclosed by faces. |
```

> Note that a `Node` in Ferrite is not the same as a `Vertex`.

Reviewer: ```suggestion
Note that a node in Ferrite is not the same as a vertex.
```

> Vertices denote endpoints of edges, while nodes may also be be located in the middle

Author: ```suggestion
Vertices denote endpoints of edges, while nodes may also be located in the middle
```

> Note that a `Node` in Ferrite is not the same as a `Vertex`.
> Vertices denote endpoints of edges, while nodes may also be be located in the middle
> of edges (e.g. for a `QuadraticLine` cell).

Reviewer 2: ```suggestion
Note that a `Node` in Ferrite is not the same as a `Vertex`.
Vertices denote endpoints of lines and the corners shapes with dimension larger than one, while nodes may also be located in the interior of edges, faces and volumes, acting as control points for the (geometric) non-linearity. Think for example about a `QuadraticLine` cell which has one node at the center of an edge and two more nodes located at the vertices/endpoints.
```

I think we can expand on this a bit, since this is really important to understand for users. I would suggest something in this direction, but it needs further refinement I think.

Reviewer: I thought that for interpolations we considered the middle point as a vertex and that node was just "points in a grid"...

Reviewer 2: It is a point in the grid because we enforce nodal interpolations for the geometry. :P

Author: I think of (maybe influenced by @termi-official) that we have two types of nodes
* Geometric nodes: Points in the grid (technically support points for the geometric interpolation).
* Algebraic nodes: Support points for nodal interpolations (e.g. Lagrange interpolations, but not Nedelec), as referred to [here](https://finite-element.github.io/L2_fespaces.html#ciarlet-s-finite-element). The coordinates of these nodes are not explicitly defined in the grid, but obtained by mapping their location in the reference point by the geometric interpolation of the cell. 

I've thought that it would make sense to put this in a separate doc. I don't want to put the emphasis here, since that is too hidden and takes away focus from the main point (dealing with reference shapes, and nodes haven't been introduced yet). 

I'll open a separate issue where we could list which terms could be defined on a "defintions" page, which would be good to have IMO.

Reviewer 2: > I think of (maybe influenced by @termi-official) that we have two types of nodes
> 
>     * Geometric nodes: Points in the grid (technically support points for the geometric interpolation).
> 
>     * Algebraic nodes: Support points for nodal interpolations (e.g. Lagrange interpolations, but not Nedelec), as referred to [here](https://finite-element.github.io/L2_fespaces.html#ciarlet-s-finite-element). The coordinates of these nodes are not explicitly defined in the grid, but obtained by mapping their location in the reference point by the geometric interpolation of the cell.
> 

*achievement unlocked* :sunglasses: 

> 
> I've thought that it would make sense to put this in a separate doc. I don't want to put the emphasis here, since that is too hidden and takes away focus from the main point (dealing with reference shapes, and nodes haven't been introduced yet).
> 
> I'll open a separate issue where we could list which terms could be defined on a "defintions" page, which would be good to have IMO.

Makes sense. Should we then just revert the change here and add to the issue that we might want to put a reference in this page to the explanation for the nodes?

Author: I think the short note here is fine for now, and once we have a good list we can shorten it more or even remove it completely and rather link to the definitions page. 

> I'll open a separate issue

#944

To write dimensionally independent code, Ferrite also denotes entities by their
[*codimension*](https://en.wikipedia.org/wiki/Codimension),
defined relative the reference shape dimension. Specifically, Ferrite has the entities

| Entity   | Description |
| :------- | :---------- |
| `Cell`   | *0-codimensional* entity, i.e. the same as the reference shape. |
| `Facet`  | *1-codimensional* entity defining the boundary of cells. |

Standard use cases mostly deal with these codimensional entities,
such as [`CellValues`](@ref) and [`FacetValues`](@ref).

!!! note "Definition of codimension"
    In Ferrite, *codimension* is defined relative to the reference dimension of the specific entity.

>     Note that other Finite Element codes may define it differently

Reviewer: ```suggestion
    Note that other finite element codes may define it differently
```

    (e.g. relative the highest reference dimension in the grid).

## Entity numbering

> Each reference shape defines the numbering of its `Vertex`, `Edge`, and `Face` entities,
> where the `Edge` and `Face` entities are defined from their `Vertex` numbers.

Reviewer: ```suggestion
Each reference shape defines the numbering of its vertices, edges, and faces entities,
where the edge and face entities are defined from their vertex numbers.

!!! note
    The numbering and identification of entities is (mostly) for internal use and typically
    not something users of Ferrite need to interact with.
```

### Example
The `RefQuadrilateral` is defined on the domain ``[-1, 1] \times [-1, 1]``
in the local ``\xi_1-\xi_2`` coordinate system.

![local element](./assets/local_element.svg)

The vertices of a `RefQuadrilateral` are then
```@example
using Ferrite #hide
Ferrite.reference_vertices(RefQuadrilateral)
```
and its edges are then defined as
```@example
using Ferrite #hide
Ferrite.reference_edges(RefQuadrilateral)
```
where the numbers refer to the vertex number.

> Finally, this reference shape is 2-dimensional, so it only has a single `face`,
> corresponding to the `Cell`,

Reviewer: ```suggestion
Finally, this reference shape is 2-dimensional, so it only has a single face,
corresponding to the cell itself,
```

```@example
using Ferrite #hide
Ferrite.reference_faces(RefQuadrilateral)
```
also defined in terms of its vertices.

> As this is a 2-dimensional reference shape, the `facets` are the `edges`, i.e.

Reviewer: ```suggestion
As this is a 2-dimensional reference shape, the facets are the edges, i.e.
```
